### PR TITLE
Added POST to track and identify methods

### PR DESCRIPTION
**Pull request by jon-batscha**

### CODE REVIEW

File: klaviyo/public.py [hunk 1]
```diff
@@ -63,8 +72,31 @@ def track(
         if ip_address:
             params['ip'] = ip_address
 
-        query_string = self._build_query_string(params, is_test)
-        return self._public_request(self.TRACK, query_string)
+        if method not in ['get','post']:
+
+            raise ValueError('method argument must be either "post" or "get"')
+
+        else:
```
Comment (smoucka) [inline klaviyo/public.py:79]: Nit: this else isn't really needed and removing it keeps things from getting too indented below

File: klaviyo/public.py [hunk 2]
```diff
@@ -63,8 +72,31 @@ def track(
         if ip_address:
             params['ip'] = ip_address
 
-        query_string = self._build_query_string(params, is_test)
-        return self._public_request(self.TRACK, query_string)
+        if method not in ['get','post']:
```
Comment (smoucka) [inline klaviyo/public.py:75]:
```suggestion
        if method.lower() not in ['get','post']:
```

File: klaviyo/public.py [hunk 3]
```diff
@@ -63,8 +72,31 @@ def track(
...
+
```
Comment (smoucka) [inline klaviyo/public.py:76]:
```suggestion
```

File: klaviyo/public.py [hunk 4]
```diff
@@ -63,8 +72,31 @@ def track(
...
+                headers = {
+                    "Accept": "text/html",
+                    "Content-Type": "application/x-www-form-urlencoded"
+                }
+
+                transformed = 'data='+quote(dumps(params))
+
+                response = request("POST", url, data=transformed, headers=headers)
+
+                return KlaviyoAPIResponse(response.status_code, response.json())
```
Comment (smoucka) [inline klaviyo/public.py:94]: the logic here is duplicated down in the identify method. Could you dry this out a bit? perhaps a public_post_request or something similar available in the api class?

File: klaviyo/public.py [hunk 5]
```diff
@@ -19,7 +25,8 @@ def track(
         customer_properties=None,
         timestamp=None,
         ip_address=None,
-        is_test=False
+        is_test=False,
+        method='get'
```
Comment (smoucka) [inline klaviyo/public.py:29]: Can you use the existing constants?

File: klaviyo/public.py [hunk 6]
```diff
@@ -63,8 +72,31 @@ def track(
...
+            if method == 'post':
+
+                url = '{}/{}'.format(KlaviyoAPI.KLAVIYO_API_SERVER,self.TRACK)
+
+                headers = {
+                    "Accept": "text/html",
+                    "Content-Type": "application/x-www-form-urlencoded"
+                }
+
+                transformed = 'data='+quote(dumps(params))
```
Comment (smoucka) [inline klaviyo/public.py:90]: Could you rename this variable to something more descriptive? `querystring` maybe? You also might be able to utilize existing methods that exist to transform the params into a query string. Take a look at [`_build_query_string`](https://github.com/klaviyo/python-klaviyo/blob/master/klaviyo/api_helper.py#L122)

**Comment by dssemmeljumpstart** [comment]
Hey @smoucka  @jon-batscha it looks like this is the officially supported library on klaviyo, and as of right now there is a backwards incompatible breaking change due to setting a URI limit on your get request before updating your python libraries!

It looks like this diff will fix that issue!

Please let me know if there is anything I can do to assist in getting this landed, as it is causing 414 errors for code that was written and working over 3 months ago.

Sincerely - a concerned Klaviyo client :)

File: klaviyo/public.py (hunk 6 above)
Comment (jon-batscha) [reply klaviyo/public.py:90]: Good call on querystring. Separately: just looked into _build_query_string, it won't work for this case, as POST does not use base64 encoding. I looked at others, and didn't find anything that could be used, but happy to update if I'm mistaken.

File: klaviyo/public.py (hunk 4 above)
Comment (jon-batscha) [reply klaviyo/public.py]: created a separate _build_query_string_post and _public_post_request functions in api_helper

File: klaviyo/public.py (hunk 6 above)
Comment (jon-batscha) [reply klaviyo/public.py]: updated: renamed to datastring (as these are not in query params), and created a new _build_data_string helper function

**Comment by dssemmeljumpstart** [comment]
Thank you @jon-batscha !!!!!

**Comment by jon-batscha** [comment]
@dssemmeljumpstart really appreciate the feedback, feel free to reach out if anything pops up

File: klaviyo/public.py [hunk 7]
```diff
@@ -63,8 +72,27 @@ def track(
...
 
-        query_string = self._build_query_string(params, is_test)
-        return self._public_request(self.TRACK, query_string)
+        if method.lower() not in ['get','post']:
+
+            raise ValueError('method argument must be either "post" or "get"')
+
+        if method == 'post':
+
+            url = '{}/{}'.format(KlaviyoAPI.KLAVIYO_API_SERVER,self.TRACK)
```
Comment (remstone7) [inline klaviyo/public.py:81]: nit: double quotes and
`url = '{}/{}'.format(KlaviyoAPI.KLAVIYO_API_SERVER, self.TRACK)`

File: klaviyo/public.py [hunk 8]
```diff
@@ -129,8 +159,28 @@ def identify(self, email=None, external_id=None, properties={}, is_test=False):
...
         }
 
-        query_string = self._build_query_string(params, is_test)
-        return self._public_request(self.IDENTIFY, query_string)
+
+        if method.lower() not in ['get','post']:
+
+            raise ValueError('method argument must be either "post" or "get"')
+
+        if method == 'post':
```
Comment (remstone7) [inline klaviyo/public.py:167]: could we not use string literals?

`if method.upper() == KlaviyoAPI.HTTP_GET`

**Comment by dssemmeljumpstart** [comment]
Hey @jon-batscha can we get an updated version here? I am looking to update our library ASAP!

https://pypi.org/project/klaviyo/#history

**Comment by jon-batscha** [comment]
@dssemmeljumpstart looking into this now

**Comment by jon-batscha** [comment]
@dssemmeljumpstart just published latest version to PyPI, please feel free to reach out with any additional feedback regarding our SDKs, we really appreciate it

**Pull request merged**